File: teoria_da_computacao/criptomoedas/java/CryptoCoin/src/Transaction.java
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.Date;

public class Transaction {
    public String transactionId;                    // ID da transação (hash)
    public PublicKey sender;                        // Chave pública do remetente
    public PublicKey recipient;                     // Chave pública do destinatário
    public float value;                             // Valor transferido
    public byte[] signature;                        // Assinatura digital
    public ArrayList<TransactionInput> inputs;      // Entradas (UTXOs gastos)
    public ArrayList<TransactionOutput> outputs;    // Saídas (UTXOs criados)
    public long timestamp;                          // Data/hora
    
    private static int sequence = 0;                // Contador para garantir IDs únicos
    
    public Transaction(PublicKey sender, PublicKey recipient, float value, 
                       ArrayList<TransactionInput> inputs) {
        this.sender = sender;
        this.recipient = recipient;
        this.value = value;
        this.inputs = inputs;
        this.outputs = new ArrayList<>();
        this.timestamp = new Date().getTime();
    }
    
    // Calcula o hash da transação
    private String calculateHash() {
        sequence++;
        return StringUtil.applySha256(
            StringUtil.getStringFromKey(sender) +
            StringUtil.getStringFromKey(recipient) +
            Float.toString(value) + sequence + timestamp
        );
    }
    
    // Gera assinatura digital
    public void generateSignature(PrivateKey privateKey) {
        String data = StringUtil.getStringFromKey(sender) + 
                      StringUtil.getStringFromKey(recipient) + 
                      Float.toString(value);
        signature = StringUtil.applyECDSASig(privateKey, data);
    }
    
    // Verifica assinatura
    public boolean verifySignature() {
        String data = StringUtil.getStringFromKey(sender) + 
                      StringUtil.getStringFromKey(recipient) + 
                      Float.toString(value);
        return StringUtil.verifyECDSASig(sender, data, signature);
    }
    
    // Processa a transação
    public boolean processTransaction() {
        // Verifica assinatura
        if (!verifySignature()) {
            System.out.println("Assinatura inválida!");
            return false;
        }
        
        // Coleta UTXOs das entradas
        for (TransactionInput input : inputs) {
            input.UTXO = Blockchain.UTXOs.get(input.transactionOutputId);
        }
        
        // Verifica se há saldo suficiente
        float totalInput = getInputsValue();
        if (totalInput < value) {
            System.out.println("Saldo insuficiente! Total disponível: " + totalInput + 
                               " - Necessário: " + value);
            return false;
        }
        
        // Remove UTXOs gastos
        for (TransactionInput input : inputs) {
            Blockchain.UTXOs.remove(input.transactionOutputId);
        }
        
        // Cria saída para o destinatário
        float leftOver = totalInput - value;
        transactionId = calculateHash();
        outputs.add(new TransactionOutput(this.recipient, value, transactionId));
        outputs.add(new TransactionOutput(this.sender, leftOver, transactionId));
        
        // Adiciona novas UTXOs
        for (TransactionOutput output : outputs) {
            Blockchain.UTXOs.put(output.id, output);
        }
        
        return true;
    }
    
    // Calcula valor total das entradas
    public float getInputsValue() {
        float total = 0;
        for (TransactionInput input : inputs) {
            if (input.UTXO != null) {
                total += input.UTXO.value;
            }
        }
        return total;
    }
    
    // Calcula valor total das saídas
    public float getOutputsValue() {
        float total = 0;
        for (TransactionOutput output : outputs) {
            total += output.value;
        }
        return total;
    }
}

File: teoria_da_computacao/criptomoedas/python/criptocoin/transaction.py
import time
from typing import List, Dict, Optional, TYPE_CHECKING
from utils import Utils
from transaction_input import TransactionInput
from transaction_output import TransactionOutput

if TYPE_CHECKING:
    from wallet import Wallet

class Transaction:
    """Transação entre carteiras"""
    
    # UTXOs globais (simulando um estado compartilhado)
    global_utxos: Dict[str, TransactionOutput] = {}
    
    def __init__(self, sender, recipient, value: float, inputs: List[TransactionInput]):
        self.transaction_id = None
        self.sender = sender          # Chave pública do remetente
        self.recipient = recipient    # Chave pública do destinatário
        self.value = value
        self.signature = None
        self.inputs = inputs if inputs else []
        self.outputs: List[TransactionOutput] = []
        self.timestamp = int(time.time() * 1000)
        self.transaction_id = self.calculate_hash()
    
    def calculate_hash(self) -> str:
        """Calcula o hash da transação"""
        from utils import Utils
        sender_str = Utils.get_string_from_key(self.sender) if self.sender else ''
        recipient_str = Utils.get_string_from_key(self.recipient) if self.recipient else ''
        return Utils.sha256(
            sender_str + recipient_str + 
            str(self.value) + str(self.timestamp)
        )
    
    def generate_signature(self, private_key) -> None:
        """Gera assinatura digital"""
        from utils import Utils
        sender_str = Utils.get_string_from_key(self.sender)
        recipient_str = Utils.get_string_from_key(self.recipient)
        data = sender_str + recipient_str + str(self.value)
        self.signature = Utils.apply_ecdsa_sig(private_key, data)
    
    def verify_signature(self) -> bool:
        """Verifica assinatura digital"""
        from utils import Utils
        if self.sender is None:
            return True  # Transação coinbase (mineração)
        
        sender_str = Utils.get_string_from_key(self.sender)
        recipient_str = Utils.get_string_from_key(self.recipient)
        data = sender_str + recipient_str + str(self.value)
        return Utils.verify_ecdsa_sig(self.sender, data, self.signature)
    
    def get_inputs_value(self) -> float:
        """Calcula valor total das entradas"""
        total = 0.0
        for inp in self.inputs:
            if inp.utxo:
                total += inp.utxo.value
        return total
    
    def get_outputs_value(self) -> float:
        """Calcula valor total das saídas"""
        return sum(out.value for out in self.outputs)
    
    def process_transaction(self) -> bool:
        """Processa a transação"""
        # Verifica assinatura
        if not self.verify_signature():
            print("Assinatura inválida!")
            return False
        
        # Coleta UTXOs das entradas
        for inp in self.inputs:
            inp.utxo = Transaction.global_utxos.get(inp.transaction_output_id)
            if inp.utxo is None:
                print(f"UTXO não encontrado: {inp.transaction_output_id}")
                return False
        
        # Verifica saldo
        total_input = self.get_inputs_value()
        if total_input < self.value:
            print(f"Saldo insuficiente! Total disponível: {total_input} - Necessário: {self.value}")
            return False
        
        # Remove UTXOs gastos
        for inp in self.inputs:
            if inp.transaction_output_id in Transaction.global_utxos:
                del Transaction.global_utxos[inp.transaction_output_id]
        
        # Cria saídas
        leftover = total_input - self.value
        self.transaction_id = self.calculate_hash()
        
        # Saída para o destinatário
        self.outputs.append(TransactionOutput(self.recipient, self.value, self.transaction_id))
        # Saída de troco para o remetente
        if leftover > 0:
            self.outputs.append(TransactionOutput(self.sender, leftover, self.transaction_id))
        
        # Adiciona novas UTXOs
        for output in self.outputs:
            Transaction.global_utxos[output.id] = output
        
        return True
    
    def to_dict(self) -> dict:
        """Converte para dicionário (para debug)"""
        from utils import Utils
        return {
            'transaction_id': self.transaction_id,
            'sender': Utils.get_string_from_key(self.sender) if self.sender else 'COINBASE',
            'recipient': Utils.get_string_from_key(self.recipient),
            'value': self.value,
            'timestamp': self.timestamp,
            'inputs': [inp.to_dict() for inp in self.inputs],
            'outputs': [out.to_dict() for out in self.outputs]
        }
